src/tileset.rs

```diff
@@ -0,0 +1,308 @@
+use crate::tileset::file_tileset::FileTileSet;
+use crate::tileset::hgt::HGT;
+use crate::tileset::http_tileset::HTTPTileSet;
+use std::collections::{HashMap, VecDeque};
+use std::time::Duration;
+use tokio::sync::{Mutex, oneshot};
+use tokio::time::timeout;
+
+mod file_tileset;
+mod hgt;
+mod http_tileset;
+
+#[derive(Debug, Clone)]
+pub struct TileSetOptions {
+    pub path: String,
+    pub cache_size: u64,
+    pub gzip: bool,
+}
+
+impl Default for TileSetOptions {
+    fn default() -> Self {
+        Self {
+            path: String::new(),
+            cache_size: 128,
+            gzip: true,
+        }
+    }
+}
+
+pub enum TileSet {
+    File(FileTileSet),
+    HTTP(HTTPTileSet),
+}
+
+impl TileSet {
+    pub fn new(options: TileSetOptions) -> Result<Self, Box<dyn std::error::Error>> {
+        if options.path.starts_with("http://") || options.path.starts_with("https://") {
+            Ok(TileSet::HTTP(HTTPTileSet::new(
+                options.path.clone(),
+                options,
+            )))
+        } else {
+            Ok(TileSet::File(FileTileSet::new(
+                options.path.clone(),
+                options,
+            )))
+        }
+    }
+}
+
+pub struct TileSetCache {
+    cache: Mutex<HashMap<(i32, i32), Vec<u8>>>,
+    order: Mutex<VecDeque<(i32, i32)>>, // Tracks the order of keys for LRU eviction
+    max_size: usize,                    // Maximum number of tiles in the cache
+    pending_fetches: Mutex<HashMap<(i32, i32), Vec<oneshot::Sender<Vec<u8>>>>>, // Track in-progress fetches
+    fetch_timeout: Duration, // Maximum time to wait for a pending fetch
+}
+
+impl TileSetCache {
```

Comment:
I just stumbled upon https://crates.io/crates/cache_loader_async, which implements an async lru cache. I think it would be useful to use a crate here instead of reimplementing it, because these concurrent caches can be quite tricky to get right and performant.

Reply:
Great!